File: ark_oauth_client_nodejs/src/jwt.js
import { constants, createPrivateKey, createPublicKey, sign as cryptoSign, verify as cryptoVerify } from 'node:crypto';
import { base64UrlDecode, base64UrlEncode, leftHalfHash } from './crypto.js';
import { ArkTokenError } from './errors.js';

/**
 * JWT decoding, signature verification and claim validation.
 *
 * The Ark server signs with RS256 and publishes its keys at `jwks_uri`, so that is the path this
 * is tuned for; the other JWS families are accepted so the same client can be pointed at Entra ID,
 * Okta or Auth0 without a second implementation.
 *
 * The one algorithm deliberately not supported is `none`. An unsigned token is a token anyone can
 * write, and every historical JWT library vulnerability of note comes from honouring the `alg`
 * header without first deciding which algorithms are acceptable.
 */

const SUPPORTED_ALGORITHMS = new Map([
  ['RS256', { hash: 'sha256', kty: 'RSA' }],
  ['RS384', { hash: 'sha384', kty: 'RSA' }],
  ['RS512', { hash: 'sha512', kty: 'RSA' }],
  ['PS256', { hash: 'sha256', kty: 'RSA', pss: true }],
  ['PS384', { hash: 'sha384', kty: 'RSA', pss: true }],
  ['PS512', { hash: 'sha512', kty: 'RSA', pss: true }],
  ['ES256', { hash: 'sha256', kty: 'EC' }],
  ['ES384', { hash: 'sha384', kty: 'EC' }],
  ['ES512', { hash: 'sha512', kty: 'EC' }]
]);

/**
 * Splits a compact JWS without verifying anything.
 *
 * Useful for reading `kid` before a key is chosen, or for logging a token's `sub` while
 * diagnosing a failure — never for deciding anything. Nothing in this library authorises on the
 * result of a decode.
 */
export function decodeJwt(token) {
  if (typeof token !== 'string') throw new ArkTokenError('the token is not a string.');
  const parts = token.split('.');
  if (parts.length !== 3) {
    throw new ArkTokenError(`the token is not a compact JWS: expected 3 segments, found ${parts.length}.`);
  }

  let header;
  let payload;
  try {
    header = JSON.parse(base64UrlDecode(parts[0]).toString('utf8'));
    payload = JSON.parse(base64UrlDecode(parts[1]).toString('utf8'));
  } catch (cause) {
    throw new ArkTokenError('the token header or payload is not valid JSON.', { cause });
  }

  return {
    header,
    payload,
    signature: base64UrlDecode(parts[2]),
    signingInput: `${parts[0]}.${parts[1]}`
  };
}

/** Verifies the signature of an already-decoded token against one JWK. */
export function verifySignature(decoded, jwk, { algorithms } = {}) {
  const alg = decoded.header.alg;
  if (!alg || alg === 'none') {
    throw new ArkTokenError(`the token declares alg '${alg ?? 'missing'}'; unsigned tokens are never accepted.`);
  }

  const allowed = algorithms ?? [...SUPPORTED_ALGORITHMS.keys()];
  if (!allowed.includes(alg)) {
    throw new ArkTokenError(`the token is signed with ${alg}, which this client does not accept (allowed: ${allowed.join(', ')}).`);
  }

  const spec = SUPPORTED_ALGORITHMS.get(alg);
  if (!spec) throw new ArkTokenError(`unsupported signing algorithm '${alg}'.`);
  if (jwk.kty !== spec.kty) {
    throw new ArkTokenError(`the signing key is a ${jwk.kty} key but the token declares ${alg}.`);
  }
  if (jwk.alg && jwk.alg !== alg) {
    throw new ArkTokenError(`the key published for kid '${jwk.kid}' is registered for ${jwk.alg}, not ${alg}.`);
  }

  let key;
  try {
    key = createPublicKey({ key: jwk, format: 'jwk' });
  } catch (cause) {
    throw new ArkTokenError(`the published key for kid '${jwk.kid}' could not be imported: ${cause.message}`, { cause });
  }

  const keyOptions = { key };
  if (spec.pss) {
    keyOptions.padding = constants.RSA_PKCS1_PSS_PADDING;
    keyOptions.saltLength = constants.RSA_PSS_SALTLEN_DIGEST;
  }
  if (spec.kty === 'EC') {
    // JWS carries an EC signature as raw r||s, not the DER sequence Node verifies by default.
    keyOptions.dsaEncoding = 'ieee-p1363';
  }

  const ok = cryptoVerify(spec.hash, Buffer.from(decoded.signingInput, 'utf8'), keyOptions, decoded.signature);
  if (!ok) throw new ArkTokenError('the token signature does not verify against the published key.');
  return true;
}

/**
 * Validates the registered claims.
 *
 * Order matters less than completeness here: skipping any single one of these is how a token
 * meant for a different application, a different provider or a different moment ends up
 * authorising a request.
 */
export function validateClaims(payload, options = {}) {
  const {
    issuer,
    audience,
    subject,
    nonce,
    maxAgeSeconds,
    clockToleranceSeconds = 60,
    requireExp = true,
    requireIat = false,
    now = Math.floor(Date.now() / 1000)
  } = options;

  const fail = (message, claim) => {
    throw new ArkTokenError(message, { claim });
  };

  if (issuer) {
    if (!payload.iss) fail('the token has no `iss` claim.', 'iss');
    if (payload.iss !== issuer) {
      fail(`the token was issued by '${payload.iss}', not by '${issuer}'.`, 'iss');
    }
  }

  if (audience) {
    const auds = Array.isArray(payload.aud) ? payload.aud : payload.aud ? [payload.aud] : [];
    if (auds.length === 0) fail('the token has no `aud` claim.', 'aud');
    if (!auds.includes(audience)) {
      fail(`the token is addressed to ${auds.map((a) => `'${a}'`).join(', ')}, not to '${audience}'.`, 'aud');
    }
    // OIDC Core §3.1.3.7: with several audiences, `azp` must name the party the token is for.
    if (auds.length > 1 && payload.azp && payload.azp !== audience) {
      fail(`the token names '${payload.azp}' as its authorized party, not '${audience}'.`, 'azp');
    }
  }

  if (subject && payload.sub !== subject) {
    fail(`the token belongs to subject '${payload.sub}', not '${subject}'.`, 'sub');
  }

  if (payload.exp === undefined) {
    if (requireExp) fail('the token has no `exp` claim.', 'exp');
  } else if (typeof payload.exp !== 'number' || now >= payload.exp + clockToleranceSeconds) {
    fail(`the token expired at ${new Date((payload.exp ?? 0) * 1000).toISOString()}.`, 'exp');
  }

  if (payload.nbf !== undefined && now + clockToleranceSeconds < payload.nbf) {
    fail(`the token is not valid before ${new Date(payload.nbf * 1000).toISOString()}.`, 'nbf');
  }

  if (payload.iat === undefined) {
    if (requireIat) fail('the token has no `iat` claim.', 'iat');
  } else if (now + clockToleranceSeconds < payload.iat) {
    fail(`the token was issued in the future, at ${new Date(payload.iat * 1000).toISOString()}.`, 'iat');
  }

  if (nonce !== undefined && nonce !== null) {
    // A missing nonce is as bad as a wrong one: it means the ID token was not bound to our
    // authorization request and could have been minted for someone else's session.
    if (!payload.nonce) fail('the ID token has no `nonce`, so it cannot be tied to this sign-in.', 'nonce');
    if (payload.nonce !== nonce) fail('the ID token `nonce` does not match the one sent with the authorization request.', 'nonce');
  }

  if (maxAgeSeconds !== undefined && maxAgeSeconds !== null) {
    if (typeof payload.auth_time !== 'number') {
      fail('`max_age` was requested but the ID token carries no `auth_time`.', 'auth_time');
    }
    if (now - payload.auth_time > maxAgeSeconds + clockToleranceSeconds) {
      fail(`the user authenticated ${now - payload.auth_time}s ago, beyond the requested max_age of ${maxAgeSeconds}s.`, 'auth_time');
    }
  }

  return payload;
}

/**
 * Checks `at_hash` / `c_hash` (OIDC Core §3.1.3.6).
 *
 * These are what stop an attacker swapping in an access token or an authorization code of their
 * own alongside a genuine ID token — the substitution attack the hashes exist for. The Ark server
 * always issues both, so a missing one is worth surfacing rather than skipping quietly.
 */
export function validateTokenHashes(payload, { accessToken, code, require: required = true } = {}) {
  const check = (claim, value) => {
    if (!value) return;
    const present = payload[claim];
    if (!present) {
      if (required) throw new ArkTokenError(`the ID token has no \`${claim}\`, so the ${claim === 'at_hash' ? 'access token' : 'authorization code'} it arrived with cannot be tied to it.`, { claim });
      return;
    }
    if (present !== leftHalfHash(value)) {
      throw new ArkTokenError(`the ID token \`${claim}\` does not cover the value it arrived with — it may have been substituted.`, { claim });
    }
  };

  check('at_hash', accessToken);
  check('c_hash', code);
  return payload;
}

/**
 * The whole check for one token: signature against the provider's published keys, then claims.
 *
 * @param {string} token
 * @param {{getSigningKey(kid: string|undefined, alg: string): Promise<object>}} jwks
 */
export async function verifyJwt(token, jwks, options = {}) {
  const decoded = decodeJwt(token);

  if (options.typ) {
    // RFC 9068 §4: an access token declares `at+jwt`, which is how a resource server refuses an
    // ID token presented in its place. Both spellings are seen in the wild.
    const typ = String(decoded.header.typ ?? '').toLowerCase();
    const expected = String(options.typ).toLowerCase();
    if (typ !== expected && typ !== `application/${expected}`) {
      throw new ArkTokenError(`expected a token of type '${options.typ}' but the header declares '${decoded.header.typ ?? 'none'}'.`);
    }
  }

  const jwk = await jwks.getSigningKey(decoded.header.kid, decoded.header.alg);
  verifySignature(decoded, jwk, { algorithms: options.algorithms });
  validateClaims(decoded.payload, options);
  if (options.accessToken || options.code) {
    validateTokenHashes(decoded.payload, {
      accessToken: options.accessToken,
      code: options.code,
      require: options.requireTokenHashes !== false
    });
  }
  return decoded.payload;
}

/**
 * Signs a compact JWS.
 *
 * Needed for exactly one thing on the client side: the `private_key_jwt` client assertion (OIDC
 * Core §9), where the client proves who it is with a signature instead of a shared secret that
 * has to be stored, rotated and kept out of logs on both ends.
 *
 * @param {object} payload
 * @param {object} options
 * @param {import('node:crypto').KeyObject|string|object} options.key PEM, JWK or KeyObject
 * @param {string} [options.alg] RS256 by default — what the Ark server accepts
 * @param {string} [options.kid] published in the header so the verifier can pick the right key
 * @param {string} [options.typ]
 */
export function signJwt(payload, { key, alg = 'RS256', kid, typ = 'JWT', header: extraHeader } = {}) {
  const spec = SUPPORTED_ALGORITHMS.get(alg);
  if (!spec) throw new ArkTokenError(`cannot sign with '${alg}'; supported: ${[...SUPPORTED_ALGORITHMS.keys()].join(', ')}.`);

  let privateKey;
  try {
    privateKey =
      typeof key === 'object' && key !== null && !Buffer.isBuffer(key) && key.asymmetricKeyType
        ? key
        : createPrivateKey(typeof key === 'object' && !Buffer.isBuffer(key) ? { key, format: 'jwk' } : key);
  } catch (cause) {
    throw new ArkTokenError(`the signing key could not be imported: ${cause.message}`, { cause });
  }

  const head = { alg, typ, ...(kid ? { kid } : {}), ...extraHeader };
  const signingInput = `${base64UrlEncode(JSON.stringify(head))}.${base64UrlEncode(JSON.stringify(payload))}`;

  const keyOptions = { key: privateKey };
  if (spec.pss) {
    keyOptions.padding = constants.RSA_PKCS1_PSS_PADDING;
    keyOptions.saltLength = constants.RSA_PSS_SALTLEN_DIGEST;
  }
  if (spec.kty === 'EC') keyOptions.dsaEncoding = 'ieee-p1363';

  const signature = cryptoSign(spec.hash, Buffer.from(signingInput, 'utf8'), keyOptions);
  return `${signingInput}.${base64UrlEncode(signature)}`;
}

File: ark_oauth_client_py/src/ark_oauth_client/jwt.py
"""
JWT decoding, signature verification and claim validation.

The Ark server signs with RS256 and publishes its keys at ``jwks_uri``, so that is the path this is
tuned for; the other JWS families are accepted so the same client can be pointed at Entra ID, Okta
or Auth0 without a second implementation.

The one algorithm deliberately not supported is ``none``. An unsigned token is a token anyone can
write, and every historical JWT library vulnerability of note comes from honouring the ``alg``
header without first deciding which algorithms are acceptable.

Signature work is delegated to :mod:`cryptography`. Verifying RS256 by hand is a page of modular
arithmetic that a library has no business shipping, and the ECDSA and PSS families are worse.
"""

from __future__ import annotations

import json
import time
from typing import Any, Dict, Mapping, NamedTuple, Optional, Sequence

from .crypto import base64url_decode, base64url_encode, left_half_hash
from .errors import ArkTokenError

__all__ = [
    "DecodedJwt",
    "SUPPORTED_ALGORITHMS",
    "decode_jwt",
    "verify_signature",
    "validate_claims",
    "validate_token_hashes",
    "verify_jwt",
    "sign_jwt",
]


class _Alg(NamedTuple):
    hash_name: str
    kty: str
    pss: bool = False


SUPPORTED_ALGORITHMS: Dict[str, _Alg] = {
    "RS256": _Alg("sha256", "RSA"),
    "RS384": _Alg("sha384", "RSA"),
    "RS512": _Alg("sha512", "RSA"),
    "PS256": _Alg("sha256", "RSA", True),
    "PS384": _Alg("sha384", "RSA", True),
    "PS512": _Alg("sha512", "RSA", True),
    "ES256": _Alg("sha256", "EC"),
    "ES384": _Alg("sha384", "EC"),
    "ES512": _Alg("sha512", "EC"),
}


class DecodedJwt(NamedTuple):
    header: Dict[str, Any]
    payload: Dict[str, Any]
    signature: bytes
    signing_input: bytes


def _cryptography():
    try:
        import cryptography  # noqa: F401
    except ImportError as cause:  # pragma: no cover - import guard
        raise ArkTokenError(
            "the 'cryptography' package is required to verify token signatures. "
            "Install it with: pip install ark-oauth-client"
        ) from cause
    return cryptography


def decode_jwt(token: str) -> DecodedJwt:
    """
    Splits a compact JWS without verifying anything.

    Useful for reading ``kid`` before a key is chosen, or for logging a token's ``sub`` while
    diagnosing a failure — never for deciding anything. Nothing in this library authorises on the
    result of a decode.
    """
    if not isinstance(token, str):
        raise ArkTokenError("the token is not a string.")
    parts = token.split(".")
    if len(parts) != 3:
        raise ArkTokenError(
            f"the token is not a compact JWS: expected 3 segments, found {len(parts)}."
        )

    try:
        header = json.loads(base64url_decode(parts[0]))
        payload = json.loads(base64url_decode(parts[1]))
    except (ValueError, TypeError) as cause:
        raise ArkTokenError("the token header or payload is not valid JSON.") from cause
    if not isinstance(header, dict) or not isinstance(payload, dict):
        raise ArkTokenError("the token header or payload is not a JSON object.")

    try:
        # Decoded here rather than at the point of verification so that a malformed token fails as
        # an ArkTokenError like every other bad token, instead of as a base64 error out of a
        # library the caller never called.
        signature = base64url_decode(parts[2])
    except (ValueError, TypeError) as cause:
        raise ArkTokenError("the token signature is not valid base64url.") from cause

    return DecodedJwt(
        header=header,
        payload=payload,
        signature=signature,
        signing_input=f"{parts[0]}.{parts[1]}".encode("ascii"),
    )


def _hash_algorithm(name: str):
    from cryptography.hazmat.primitives import hashes

    return {"sha256": hashes.SHA256(), "sha384": hashes.SHA384(), "sha512": hashes.SHA512()}[name]


def _public_key_from_jwk(jwk: Mapping[str, Any]):
    """Imports a JWK as a cryptography public key object."""
    from cryptography.hazmat.primitives.asymmetric import ec, rsa

    def num(name: str) -> int:
        return int.from_bytes(base64url_decode(jwk[name]), "big")

    kty = jwk.get("kty")
    if kty == "RSA":
        return rsa.RSAPublicNumbers(e=num("e"), n=num("n")).public_key()
    if kty == "EC":
        curves = {"P-256": ec.SECP256R1(), "P-384": ec.SECP384R1(), "P-521": ec.SECP521R1()}
        curve = curves.get(str(jwk.get("crv")))
        if curve is None:
            raise ArkTokenError(f"the published key uses curve '{jwk.get('crv')}', which is not supported.")
        return ec.EllipticCurvePublicNumbers(x=num("x"), y=num("y"), curve=curve).public_key()
    raise ArkTokenError(f"the published key is of type '{kty}', which is not supported.")


def verify_signature(
    decoded: DecodedJwt,
    jwk: Mapping[str, Any],
    *,
    algorithms: Optional[Sequence[str]] = None,
) -> bool:
    """Verifies the signature of an already-decoded token against one JWK."""
    _cryptography()
    from cryptography.exceptions import InvalidSignature
    from cryptography.hazmat.primitives.asymmetric import ec, padding, utils

    alg = decoded.header.get("alg")
    if not alg or alg == "none":
        raise ArkTokenError(
            f"the token declares alg '{alg or 'missing'}'; unsigned tokens are never accepted."
        )

    allowed = list(algorithms) if algorithms else list(SUPPORTED_ALGORITHMS)
    if alg not in allowed:
        raise ArkTokenError(
            f"the token is signed with {alg}, which this client does not accept "
            f"(allowed: {', '.join(allowed)})."
        )

    spec = SUPPORTED_ALGORITHMS.get(alg)
    if spec is None:
        raise ArkTokenError(f"unsupported signing algorithm '{alg}'.")
    if jwk.get("kty") != spec.kty:
        raise ArkTokenError(
            f"the signing key is a {jwk.get('kty')} key but the token declares {alg}."
        )
    if jwk.get("alg") and jwk.get("alg") != alg:
        raise ArkTokenError(
            f"the key published for kid '{jwk.get('kid')}' is registered for {jwk.get('alg')}, not {alg}."
        )

    try:
        key = _public_key_from_jwk(jwk)
    except ArkTokenError:
        raise
    except Exception as cause:
        raise ArkTokenError(
            f"the published key for kid '{jwk.get('kid')}' could not be imported: {cause}"
        ) from cause

    digest = _hash_algorithm(spec.hash_name)
    try:
        if spec.kty == "EC":
            # JWS carries an EC signature as raw r||s, not the DER sequence cryptography verifies.
            half = len(decoded.signature) // 2
            der = utils.encode_dss_signature(
                int.from_bytes(decoded.signature[:half], "big"),
                int.from_bytes(decoded.signature[half:], "big"),
            )
            key.verify(der, decoded.signing_input, ec.ECDSA(digest))
        elif spec.pss:
            key.verify(
                decoded.signature,
                decoded.signing_input,
                padding.PSS(mgf=padding.MGF1(digest), salt_length=digest.digest_size),
                digest,
            )
        else:
            key.verify(decoded.signature, decoded.signing_input, padding.PKCS1v15(), digest)
    except InvalidSignature as cause:
        raise ArkTokenError("the token signature does not verify against the published key.") from cause

    return True


def validate_claims(
    payload: Mapping[str, Any],
    *,
    issuer: Optional[str] = None,
    audience: Optional[str] = None,
    subject: Optional[str] = None,
    nonce: Optional[str] = None,
    max_age_seconds: Optional[int] = None,
    clock_tolerance_seconds: int = 60,
    require_exp: bool = True,
    require_iat: bool = False,
    now: Optional[int] = None,
    **_ignored: Any,
) -> Mapping[str, Any]:
    """
    Validates the registered claims.

    Order matters less than completeness here: skipping any single one of these is how a token
    meant for a different application, a different provider or a different moment ends up
    authorising a request.
    """
    now = int(time.time()) if now is None else now

    def fail(message: str, claim: str) -> None:
        raise ArkTokenError(message, claim=claim)

    def moment(value: Any) -> str:
        try:
            return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(float(value)))
        except (TypeError, ValueError):
            return str(value)

    if issuer:
        if not payload.get("iss"):
            fail("the token has no `iss` claim.", "iss")
        if payload.get("iss") != issuer:
            fail(f"the token was issued by '{payload.get('iss')}', not by '{issuer}'.", "iss")

    if audience:
        raw_aud = payload.get("aud")
        auds = list(raw_aud) if isinstance(raw_aud, (list, tuple)) else ([raw_aud] if raw_aud else [])
        if not auds:
            fail("the token has no `aud` claim.", "aud")
        if audience not in auds:
            listed = ", ".join(f"'{a}'" for a in auds)
            fail(f"the token is addressed to {listed}, not to '{audience}'.", "aud")
        # OIDC Core §3.1.3.7: with several audiences, `azp` must name the party the token is for.
        if len(auds) > 1 and payload.get("azp") and payload.get("azp") != audience:
            fail(
                f"the token names '{payload.get('azp')}' as its authorized party, not '{audience}'.",
                "azp",
            )

    if subject and payload.get("sub") != subject:
        fail(f"the token belongs to subject '{payload.get('sub')}', not '{subject}'.", "sub")

    exp = payload.get("exp")
    if exp is None:
        if require_exp:
            fail("the token has no `exp` claim.", "exp")
    elif not isinstance(exp, (int, float)) or now >= exp + clock_tolerance_seconds:
        fail(f"the token expired at {moment(exp)}.", "exp")

    nbf = payload.get("nbf")
    if nbf is not None and now + clock_tolerance_seconds < nbf:
        fail(f"the token is not valid before {moment(nbf)}.", "nbf")

    iat = payload.get("iat")
    if iat is None:
        if require_iat:
            fail("the token has no `iat` claim.", "iat")
    elif now + clock_tolerance_seconds < iat:
        fail(f"the token was issued in the future, at {moment(iat)}.", "iat")

    if nonce is not None:
        # A missing nonce is as bad as a wrong one: it means the ID token was not bound to our
        # authorization request and could have been minted for someone else's session.
        if not payload.get("nonce"):
            fail("the ID token has no `nonce`, so it cannot be tied to this sign-in.", "nonce")
        if payload.get("nonce") != nonce:
            fail(
                "the ID token `nonce` does not match the one sent with the authorization request.",
                "nonce",
            )

    if max_age_seconds is not None:
        auth_time = payload.get("auth_time")
        if not isinstance(auth_time, (int, float)):
            fail("`max_age` was requested but the ID token carries no `auth_time`.", "auth_time")
        elif now - auth_time > max_age_seconds + clock_tolerance_seconds:
            fail(
                f"the user authenticated {int(now - auth_time)}s ago, beyond the requested "
                f"max_age of {max_age_seconds}s.",
                "auth_time",
            )

    return payload


def validate_token_hashes(
    payload: Mapping[str, Any],
    *,
    access_token: Optional[str] = None,
    code: Optional[str] = None,
    require: bool = True,
) -> Mapping[str, Any]:
    """
    Checks ``at_hash`` / ``c_hash`` (OIDC Core §3.1.3.6).

    These are what stop an attacker swapping in an access token or an authorization code of their
    own alongside a genuine ID token — the substitution attack the hashes exist for. The Ark server
    always issues both, so a missing one is worth surfacing rather than skipping quietly.
    """

    def check(claim: str, value: Optional[str]) -> None:
        if not value:
            return
        present = payload.get(claim)
        if not present:
            if require:
                thing = "access token" if claim == "at_hash" else "authorization code"
                raise ArkTokenError(
                    f"the ID token has no `{claim}`, so the {thing} it arrived with cannot be "
                    "tied to it.",
                    claim=claim,
                )
            return
        if present != left_half_hash(value):
            raise ArkTokenError(
                f"the ID token `{claim}` does not cover the value it arrived with — it may have "
                "been substituted.",
                claim=claim,
            )

    check("at_hash", access_token)
    check("c_hash", code)
    return payload


def verify_jwt(token: str, jwks: Any, **options: Any) -> Dict[str, Any]:
    """
    The whole check for one token: signature against the provider's published keys, then claims.

    ``jwks`` is anything with ``get_signing_key(kid, alg)`` — normally a
    :class:`~ark_oauth_client.jwks.JwksCache`.
    """
    decoded = decode_jwt(token)

    typ_expected = options.get("typ")
    if typ_expected:
        # RFC 9068 §4: an access token declares `at+jwt`, which is how a resource server refuses an
        # ID token presented in its place. Both spellings are seen in the wild.
        typ = str(decoded.header.get("typ") or "").lower()
        expected = str(typ_expected).lower()
        if typ != expected and typ != f"application/{expected}":
            raise ArkTokenError(
                f"expected a token of type '{typ_expected}' but the header declares "
                f"'{decoded.header.get('typ') or 'none'}'."
            )

    jwk = jwks.get_signing_key(decoded.header.get("kid"), decoded.header.get("alg"))
    verify_signature(decoded, jwk, algorithms=options.get("algorithms"))
    validate_claims(decoded.payload, **options)
    if options.get("access_token") or options.get("code"):
        validate_token_hashes(
            decoded.payload,
            access_token=options.get("access_token"),
            code=options.get("code"),
            require=options.get("require_token_hashes", True),
        )
    return decoded.payload


def sign_jwt(
    payload: Mapping[str, Any],
    *,
    key: Any,
    alg: str = "RS256",
    kid: Optional[str] = None,
    typ: str = "JWT",
    header: Optional[Mapping[str, Any]] = None,
) -> str:
    """
    Signs a compact JWS.

    Needed for exactly one thing on the client side: the ``private_key_jwt`` client assertion (OIDC
    Core §9), where the client proves who it is with a signature instead of a shared secret that has
    to be stored, rotated and kept out of logs on both ends.

    ``key`` may be a PEM string or bytes, a JWK dict, or a :mod:`cryptography` private key object.
    """
    _cryptography()
    from cryptography.hazmat.primitives.asymmetric import ec, padding

    spec = SUPPORTED_ALGORITHMS.get(alg)
    if spec is None:
        raise ArkTokenError(
            f"cannot sign with '{alg}'; supported: {', '.join(SUPPORTED_ALGORITHMS)}."
        )

    try:
        private_key = _load_private_key(key)
    except ArkTokenError:
        raise
    except Exception as cause:
        raise ArkTokenError(f"the signing key could not be imported: {cause}") from cause

    head: Dict[str, Any] = {"alg": alg, "typ": typ}
    if kid:
        head["kid"] = kid
    head.update(header or {})

    segments = (
        base64url_encode(json.dumps(head, separators=(",", ":")))
        + "."
        + base64url_encode(json.dumps(dict(payload), separators=(",", ":")))
    )
    signing_input = segments.encode("ascii")
    digest = _hash_algorithm(spec.hash_name)

    if spec.kty == "EC":
        from cryptography.hazmat.primitives.asymmetric import utils

        der = private_key.sign(signing_input, ec.ECDSA(digest))
        r, s = utils.decode_dss_signature(der)
        size = (private_key.curve.key_size + 7) // 8
        signature = r.to_bytes(size, "big") + s.to_bytes(size, "big")
    elif spec.pss:
        signature = private_key.sign(
            signing_input,
            padding.PSS(mgf=padding.MGF1(digest), salt_length=digest.digest_size),
            digest,
        )
    else:
        signature = private_key.sign(signing_input, padding.PKCS1v15(), digest)

    return f"{segments}.{base64url_encode(signature)}"


def _load_private_key(key: Any):
    """Accepts a PEM string/bytes, a JWK mapping, or an already-loaded cryptography key."""
    from cryptography.hazmat.primitives import serialization

    if hasattr(key, "sign") and hasattr(key, "public_key"):
        return key
    if isinstance(key, Mapping):
        return _private_key_from_jwk(key)
    if isinstance(key, str):
        key = key.encode("utf-8")
    return serialization.load_pem_private_key(key, password=None)


def _private_key_from_jwk(jwk: Mapping[str, Any]):
    from cryptography.hazmat.primitives.asymmetric import ec, rsa

    def num(name: str) -> int:
        return int.from_bytes(base64url_decode(jwk[name]), "big")

    kty = jwk.get("kty")
    if kty == "RSA":
        public = rsa.RSAPublicNumbers(e=num("e"), n=num("n"))
        p, q, d = num("p"), num("q"), num("d")
        return rsa.RSAPrivateNumbers(
            p=p,
            q=q,
            d=d,
            # The CRT parameters are optional in a JWK (RFC 7518 §6.3.2); derive them when absent.
            dmp1=num("dp") if "dp" in jwk else rsa.rsa_crt_dmp1(d, p),
            dmq1=num("dq") if "dq" in jwk else rsa.rsa_crt_dmq1(d, q),
            iqmp=num("qi") if "qi" in jwk else rsa.rsa_crt_iqmp(p, q),
            public_numbers=public,
        ).private_key()
    if kty == "EC":
        curves = {"P-256": ec.SECP256R1(), "P-384": ec.SECP384R1(), "P-521": ec.SECP521R1()}
        curve = curves.get(str(jwk.get("crv")))
        if curve is None:
            raise ArkTokenError(f"the signing key uses curve '{jwk.get('crv')}', which is not supported.")
        return ec.derive_private_key(num("d"), curve)
    raise ArkTokenError(f"the signing key is of type '{kty}', which is not supported.")
